mtp_api/apps/performance/models.py

```diff
@@ -179,12 +209,30 @@ class UserSatisfaction(models.Model):
     rated_4 = models.PositiveIntegerField(verbose_name=_('Satisfied'))
     rated_5 = models.PositiveIntegerField(verbose_name=_('Very satisfied'))
 
+    objects = models.Manager.from_queryset(UserSatisfactionQueryset)()
+
     reports_start = datetime.date(2016, 11, 15)
 
     class Meta:
         ordering = ('date',)
         get_latest_by = 'date'
 
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self._percentage_satisfied = NotImplemented
+
     @property
     def all_ratings(self):
         return self.rated_1, self.rated_2, self.rated_3, self.rated_4, self.rated_5
+
+    @property
+    def percentage_satisfied(self):
+        if self._percentage_satisfied is NotImplemented:
+            count = sum(self.all_ratings)
+            self._percentage_satisfied = (self.rated_4 + self.rated_5) / count if count > 0 else None
+        return self._percentage_satisfied
+
+    @percentage_satisfied.setter
+    def percentage_satisfied(self, value):
+        # to support pre-calculating percentage on the queryset
```

Comment:
i followed the pattern used by the `DigitalTakeup` model. the property method calculates the percentage on the fly if called on an individual model but the setter is used if the `percentage_satisfied` was called on the queryset first.

```python
# uses property method
DigitalTakeup.objects.get(date=…).percentage_satisfied

# uses setter because the queryset pre-calculated the value in SQL
DigitalTakeup.objects.filter(date__gte=…).percentage_satisfied().first().percentage_satisfied
#                                         ↑ precalculates value
#                                                                ↑ creates object calling setter
#                                                                        ↑ calls getter with no need to calculate value
```

Reply:
Oh I'm glad I asked!

I completely missed this (even despite the comment). I didn't realise that the `annotate(percentage_satisfied=...` in the `QuerySet` would call this setter.

Thanks for the clarification ☺️